### PR TITLE
Refactor SubNav component

### PR DESCRIPTION
**Pull request by liywjl**
## What does this change?
Refactor SubNav component

## Why?
We plan on removing unused NAV keys from performance optimisations and a refactor of SubNav would facilitate this as it is the only component that uses this key

### CODE REVIEW

File: src/web/components/SubNav/SubNav.tsx [hunk 1]
```diff
@@ -1,87 +1,207 @@
...
+import { textSans } from '@guardian/src-foundations/typography';
+import { from } from '@guardian/src-foundations/mq';
+import { pillarPalette, pillarMap } from '@root/src/lib/pillars';
 
-import { Inner } from './Inner';
-
-interface Props {
-    subnav?: SubNavType;
+type Props = {
+    subNavSections?: SubNavType;
```
Comment (oliverlloyd) [inline src/web/components/SubNav/SubNav.tsx]: Thanks for using the same name as the property!

This prop no longer needs to be optional as there's now a guard for it in App

File: src/web/components/SubNav/SubNav.tsx [hunk 2]
```diff
@@ -1,87 +1,207 @@
-import React, { Component, createRef } from 'react';
+import React, { useState, useRef, useEffect } from 'react';
+import { css, cx } from 'emotion';
 
 import { GuardianLines } from '@frontend/web/components/GuardianLines';
+import { text } from '@guardian/src-foundations/palette';
+import { palette } from '@guardian/src-foundations';
```
Comment (oliverlloyd) [inline src/web/components/SubNav/SubNav.tsx]: I know this is just copied code but it'd be great to fix these direct uses of `palette` as we see them 🙏

File: src/web/components/SubNav/SubNav.tsx [hunk 3]
```diff
@@ -1,87 +1,207 @@
...
-    private ulRef: React.RefObject<HTMLUListElement>;
-
-    constructor(props: Props) {
-        super(props);
-        this.state = { showMore: false, isExpanded: false };
-        // More should not be shown without JS
-        this.ulRef = createRef();
+`;
+
+const subnav = css`
```
Comment (oliverlloyd) [inline src/web/components/SubNav/SubNav.tsx]: Is there a more descriptive name we can call this?

File: src/web/components/SubNav/SubNav.tsx [hunk 4]
```diff
@@ -1,87 +1,207 @@
...
-        }));
-    };
-
-    public shouldShowMore() {
-        // get ul ref
-        const ul = this.ulRef.current;
-        if (!ul) {
-            return false;
-        }
+const subnavExpanded = css`
```
Comment (oliverlloyd) [inline src/web/components/SubNav/SubNav.tsx]: Given we're in the SubNav component the subnav... prefix is redundant. Maybe `expandedStyles`?

File: src/web/components/SubNav/SubNav.tsx [hunk 5]
```diff
@@ -1,87 +1,207 @@
...
+    ${subnav};
+`;
+
+const subnavCollapsed = css`
```
Comment (oliverlloyd) [inline src/web/components/SubNav/SubNav.tsx]: As above.

File: src/web/components/SubNav/SubNav.tsx [hunk 6]
```diff
@@ -1,87 +1,207 @@
...
+    padding: 0 5px;
+    height: 36px;
+    /* Design System: Line height is being used here for centering layout, we need the primitives */
+    line-height: 36px;
 
-        return ulTop !== liTop;
+    ${from.tablet} {
+        height: 42px;
+        /* Design System: Line height is being used here for centering layout, we need the primitives */
+        /* stylelint-disable-next-line property-blacklist */
```
Comment (oliverlloyd) [inline src/web/components/SubNav/SubNav.tsx]: I'm not certain but I think we don't need this anymore

File: src/web/components/SubNav/SubNav.tsx [hunk 7]
```diff
@@ -1,87 +1,207 @@
...
+    ${from.desktop} {
+        display: none;
+    }
+`;
+
+const parentLinkStyle = css`
+    ${linkStyle};
+    font-weight: 700;
+`;
+const ps1 = css`
```
Comment (oliverlloyd) [inline src/web/components/SubNav/SubNav.tsx]: Coming at this cold, I'm at a loss what ps1 means. I know these are existing names but it'd be great to include some more meaningful ones as part of the refactor

File: src/web/components/SubNav/SubNav.tsx [hunk 8]
```diff
@@ -1,87 +1,207 @@
...
+                        className={moreStyle}
+                    >
+                        {isExpanded ? 'Less' : 'More'}
+                    </button>
+                )}
             </div>
-        );
-    }
-}
+            <GuardianLines pillar={pillar} />
```
Comment (oliverlloyd) [inline src/web/components/SubNav/SubNav.tsx]: Hello, what's this doing here? This component appears to be lost and should be gently led to the layouts file

File: src/web/layouts/CommentLayout.tsx [hunk 9]
```diff
@@ -245,7 +245,7 @@ export const CommentLayout = ({ CAPI, NAV }: Props) => {
                     islandId="sub-nav-root"
                 >
                     <SubNav
-                        subnav={NAV.subNavSections}
+                        subNavSections={NAV.subNavSections}
                         currentNavLink={NAV.currentNavLink}
                         pillar={CAPI.pillar}
```
Comment (oliverlloyd) [inline src/web/layouts/CommentLayout.tsx]: This makes so much more sense 👍

File: src/web/browser/App.tsx [hunk 10]
```diff
@@ -124,13 +124,15 @@ const App = ({ CAPI, NAV }: Props) => {
...
-            </Hydrate>
+            {NAV.subNavSections && (
+                <Hydrate root="sub-nav-root">
+                    <SubNav
+                        subNavSections={NAV.subNavSections}
+                        currentNavLink={NAV.currentNavLink}
+                        pillar={CAPI.pillar}
+                    />
+                </Hydrate>
+            )}
```
Comment (oliverlloyd) [inline src/web/browser/App.tsx]: 💯 This component would never have rendered on the server if it has no subNavSections so trying to hydrate it here was a waste, thanks for fixing!

File: src/web/components/SubNav/SubNav.tsx (hunk 4 above)
Comment (oliverlloyd) [reply src/web/components/SubNav/SubNav.tsx]: Also, we rarely use this composition method (this is the first time I've seen it in DCR) - it'd be great to refactor this to use `cx` instead for consistency

File: src/web/components/SubNav/SubNav.tsx [hunk 11]
```diff
@@ -1,87 +1,207 @@
...
+    useEffect(() => {
+        const ulEl = ulRef.current;
+        if (ulEl) {
+            const lis = ulEl.querySelectorAll('li');
+            const lastLi = lis[lis.length - 1];
+
+            const ulTop = ulEl.getBoundingClientRect().top;
+            const liTop = lastLi.getBoundingClientRect().top;
+
+            setShowMore(ulTop !== liTop);
```
Comment (oliverlloyd) [inline src/web/components/SubNav/SubNav.tsx]: I was initially confused what this was doing but read through it and it's actually quite a neat solution. It's a shame to have 2 `getBoundingClientRect` calls though, they're pretty expensive. I honestly can't think of a way off the top of my head but if while you're in the code you can see a way to refactor this to a more optimal method... Maybe `offsetTop`?

File: src/web/components/SubNav/SubNav.tsx (hunk 8 above)
Comment (liywjl) [reply src/web/components/SubNav/SubNav.tsx]: good point!!

File: src/web/components/SubNav/SubNav.tsx (hunk 7 above)
Comment (liywjl) [reply src/web/components/SubNav/SubNav.tsx]: Yeah idk what this is meant to be referring to, I will try find a better name

File: src/web/components/SubNav/SubNav.tsx (hunk 11 above)
Comment (liywjl) [reply src/web/components/SubNav/SubNav.tsx]: thanks for pointing it out, I will try and find a cleaner solution

Comment (liywjl) [reply src/web/components/SubNav/SubNav.tsx]: after looking at it I cant think of a better solution so I am keeping it as is

File: src/web/browser/App.tsx [hunk 12]
```diff
@@ -124,13 +125,18 @@ const App = ({ CAPI, NAV }: Props) => {
...
-            </Hydrate>
+            {NAV.subNavSections && (
+                <Hydrate root="sub-nav-root">
+                    <>
+                        <SubNav
+                            subNavSections={NAV.subNavSections}
+                            currentNavLink={NAV.currentNavLink}
+                            pillar={CAPI.pillar}
+                        />
+                        <GuardianLines pillar={CAPI.pillar} />
```
Comment (oliverlloyd) [inline src/web/browser/App.tsx]: These don't need hydrating - we only need to server side render them and move on

File: src/web/layouts/CommentLayout.tsx [hunk 13]
```diff
@@ -249,6 +249,7 @@ export const CommentLayout = ({ CAPI, NAV }: Props) => {
...
                     />
+                    <GuardianLines pillar={CAPI.pillar} />
```
Comment (oliverlloyd) [inline src/web/layouts/CommentLayout.tsx]: We can drop these out to their own section - they're not really part of SubNav at all and then we have the bonus that they don't need to be rehydrated on the client

**Pull request merged**